pkg/agent/flowexporter/connections/connection_store.go

```diff
@@ -0,0 +1,513 @@
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case denyConn := <-s.denyConnUpdates:
+			s.handleDenyConnection(denyConn)
+		case <-staleConnTicker.C():
+			s.removeStaleConnections()
+		case <-pollTicker.C():
+			_, err := s.PollConntrackAndStore()
+			if err != nil {
+				// Not failing here as errors can be transient and could be resolved in future poll cycles.
+				// TODO: Come up with a backoff/retry mechanism by increasing poll interval and adding retry timeout
+				klog.Errorf("Error during conntrack poll cycle: %v", err)
+			}
+		}
+	}
+}
+
+func (s *ConnStore) handleDenyConnection(denyConn *connection.Connection) {
+	s.updateConns([]*connection.Connection{denyConn}, nil, s.denyConnectionAugment, denyConnMerge)
+}
+
+func (s *ConnStore) ctConnectionAugment(conn *connection.Connection) *connection.Connection {
+	conn = s.fillPodInfo(conn)
+	conn = s.fillServiceInfo(conn)
+	conn = s.fillNetworkPolicyMetadataInfo(conn)
+	conn = s.fillFlowType(conn)
+	conn = s.fillEgressInfo(conn)
+
+	return conn
+}
+
+func (s *ConnStore) denyConnectionAugment(conn *connection.Connection) *connection.Connection {
+	if conn.StartTime.IsZero() {
+		conn.StartTime = s.clock.Now()
+		conn.StopTime = conn.StartTime
+	}
+
+	conn = s.fillPodInfo(conn)
+	conn = s.fillServiceInfo(conn)
+	conn = s.fillFlowType(conn)
+	conn = s.fillNetworkPolicyMetadataInfo(conn)
+	return conn
+}
+
+func (s *ConnStore) updateConns(conns []*connection.Connection, l7Events map[connection.ConnectionKey]L7ProtocolFields, augment AugmentFn, mergeFn MergeFunc) {
+	if len(conns) == 0 {
+		return
+	}
+
+	updatedConns := make([]*connection.Connection, 0, len(conns)+len(l7Events))
+	updatedConns = s.getL7Conns(updatedConns, l7Events)
+	now := s.clock.Now()
+
+	for _, conn := range conns {
+		if conn == nil {
+			continue
+		}
+
+		key := connection.NewConnectionKey(conn)
+		existing, connExists := s.entries[key]
+
+		// This is only needed because of how "IsConnectionDying" works. We should rely on either
+		// stale connections cleanup or updated connection id in the case of conntrack flows instead.
+		// TODO: Remove usage of IsPresent
+		conn.IsPresent = !conn.IsDenyFlow
+		conn.LastUpdateTime = now
+
+		// Handle a special case where the keys are the same (same src/dst address, port and protocol)
+		// however the connection itself is new. This can happen for example when doing the following:
+		// curl --local-port 55555 <target>
+		// When this case applies, we want to re-populate the connection again since a network policy
+		// or target service may have changed while ending up with the same resulting destination pod.
+		// This applies only to CT flows; denied connections will increment the stats until it has
+		// been removed (stale) since denied connections do not specify ID.
+		// Should we consider whether the new connection is started after the existing connection?
+		if connExists && conn.ID != existing.ID {
+			existing = nil
+			connExists = false
+		}
```

Comment:
would you consider this a bug fix? If yes, I would look into addressing this separately, not as party of this PR
what does it look like from the perspective of the consumer (destination)? is there a way to distinguish between the 2 connections?